components/calendar/src/dangi.rs

```diff
@@ -0,0 +1,1221 @@
+        Some(AnyCalendarKind::Dangi)
+    }
+}
+
+impl Date<Dangi> {
+    /// Construct a new Dangi date from a `year`, `month`, and `day`.
+    /// `year` represents the Chinese year counted infinitely with -2332 (2333 BCE) as year 1;
+    /// `month` represents the month of the year ordinally (ex. if it is a leap year, the last month will be 13, not 12);
+    /// `day` indicates day of month.
+    ///
+    /// ```rust
+    /// use icu::calendar::Date;
+    ///
+    /// let date_dangi = Date::try_new_dangi_date(4356, 6, 18)
+    ///     .expect("Failed to initialize Dangi Date instance.");
+    ///
+    /// assert_eq!(date_dangi.year().number, 4356);
+    /// assert_eq!(date_dangi.year().cyclic, Some(40));
+    /// assert_eq!(date_dangi.year().related_iso, Some(2023));
+    /// assert_eq!(date_dangi.month().ordinal, 6);
+    /// assert_eq!(date_dangi.day_of_month().0, 18);
+    /// ```
+    pub fn try_new_dangi_date(year: i32, month: u8, day: u8) -> Result<Date<Dangi>, CalendarError> {
+        ArithmeticDate::new_from_lunar_ordinals(year, month, day)
+            .map(ChineseBasedDateInner)
+            .map(DangiDateInner)
+            .map(|inner| Date::from_raw(inner, Dangi))
+    }
+}
+
+impl DateTime<Dangi> {
+    /// Construct a new Dangi DateTime from integers. See [`try_new_dangi_date`].
+    ///
+    /// ```rust
+    /// use icu::calendar::DateTime;
+    ///
+    /// let dangi_datetime = DateTime::try_new_dangi_datetime(4356, 6, 6, 13, 1, 0)
+    ///     .expect("Failed to initialize Dangi DateTime instance.");
+    ///
+    /// assert_eq!(dangi_datetime.date.year().number, 4356);
+    /// assert_eq!(dangi_datetime.date.year().related_iso, Some(2023));
+    /// assert_eq!(dangi_datetime.date.year().cyclic, Some(40));
+    /// assert_eq!(dangi_datetime.date.month().ordinal, 6);
+    /// assert_eq!(dangi_datetime.date.day_of_month().0, 6);
+    /// assert_eq!(dangi_datetime.time.hour.number(), 13);
+    /// assert_eq!(dangi_datetime.time.minute.number(), 1);
+    /// assert_eq!(dangi_datetime.time.second.number(), 0);
+    /// ```
+    pub fn try_new_dangi_datetime(
+        year: i32,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<DateTime<Dangi>, CalendarError> {
+        Ok(DateTime {
+            date: Date::try_new_dangi_date(year, month, day)?,
+            time: types::Time::try_new(hour, minute, second, 0)?,
+        })
+    }
+}
+
+impl ChineseBased for Dangi {
+    fn location(fixed: RataDie) -> Location {
+        if fixed < FIXED_1908 {
+            KOREAN_LOCATION_ORIGINAL
+        } else if fixed < FIXED_1912 {
+            KOREAN_LOCATION_1908
+        } else if fixed < FIXED_1954 {
+            KOREAN_LOCATION_1912
+        } else if fixed < FIXED_1961 {
+            KOREAN_LOCATION_1954
+        } else {
+            KOREAN_LOCATION_1961
+        }
+    }
+
+    const EPOCH: RataDie = KOREAN_EPOCH;
+
+    #[allow(clippy::unwrap_used)]
```

Comment:
Done both

Reply:
Nit: You're not unwrapping any more; you're calling `new_unchecked`. However, it it clear that this is okay? We shouldn't allow any code paths where ArithmeticDate can be built with out-of-range values

Reply:
I've updated the documentation and removed the clippy tag; I think Manish and I decided it's fine to use `new_unchecked` in this context since this function is only ever called by `chinese_based_date_from_fixed`, which will generate the year, month, and day values.